curriculum/challenges/english/25-front-end-development/lecture-working-with-data-fetching-and-memoization-in-react/67d2f4ddb4a4306fdf5bbaee.md

```diff
@@ -2,13 +2,138 @@
+
+As you might guess, both `useMemo` and `useCallback` are hooks, but `React.memo` is a component wrapper, a higher-order function (HOC).
+
+In the next lecture, we will take a look at how the `useCallback` hook and `React.memo` work.
+
+`useMemo` lets you memoize computed values while `useCallback` does the same for function references.
+
+If you're wondering what computed values and function references are, computed values refer to the result of executing a function, while function references are the pointers to functions – the function object in memory.
+
+Let's see how to use the `useMemo` hook first. Here's the basic syntax of the `useMemo` hook:
+
+```js
+const memoizedValue = useMemo(
+ function () {
+   return computeExpensiveValue(a, b);
+ },
+ [a, b]
+);
+```
+
+You can see all that's needed is to wrap the `useMemo` hook around the function.
+
+This `ExpensiveSquare` component will receive a `num` prop which it will use to calculate the square:
+
+```js
+function ExpensiveSquare({ num }) {
+ function calculateSquare(n) {
+   console.log("Calculating square...");
+   return n * n;
+ }
+
+ const squared = calculateSquare(num);
+ return (
+   <p>
+     Square of {num}: {squared}
+   </p>
+ );
+}
+export default ExpensiveSquare;
+```
+
+Here's the `App` component where the `ExpensiveSquare` is being used:
+
+```js
+import { useState, useEffect } from "react";
+import ExpensiveSquare from "./components/ExpensiveSquare";
+
+function App() {
+ const [timer, setTimer] = useState(0);
+ const [num, setNum] = useState(0);
+
+ useEffect(() => {
+   const interval = setInterval(() => setTimer((c) => c + 1), 1000);
+   return () => clearInterval(interval);
+ }, []);
+
+ return (
+   <div>
+     <h1>Timer: {timer} seconds gone</h1>
+     <ExpensiveSquare num={num} />
+     <button onClick={() => setNum((n) => n + 1)}>Increase Number</button>
+   </div>
+ );
+}
+
+export default App;
+```
+
+The `timer` in the `useEffect`, running every second, will make the `calculateSquare` function runs any time it runs, even when you don't increase the `num` state variable.
+
+To solve this problem, we can use the `useMemo` hook by wrapping the function call in it and specifying the `num` variable as the dependency:
+
+```js
+// import the useMemo hook
+import { useMemo } from "react";
+
+function ExpensiveSquare({ num }) {
+  function calculateSquare(n) {
+    console.log("Calculating square...");
+    return n * n;
+  }
+
+  // const squared = calculateSquare(num);
+  // Wrap the function call in useMemo instead
+  const squared = useMemo(() => calculateSquare(num), [num]);
+
+  return (
+    <p>
+      Square of {num}: {squared}
+    </p>
+  );
+}
+
+export default ExpensiveSquare;
+```
+
+This will make sure the function is memoized by caching the result, so calculation happens only when `num` change, not when anything changes in the component it's being used in.
```

Comment:
```suggestion
This will make sure the function is memoized by caching the result, so calculation happens only when the `num` variable changes, not when anything changes in the component it's being used in.
```